easybuild/tools/build_from_easystack.py

```diff
@@ -0,0 +1,173 @@
+        self.toolchain_version = toolchain_version
+        self.toolchain_name = toolchain_name
+        self.toolchain = toolchain
+        self.versionsuffix = None
+
+    # to be implemented
+    # def get_versionsuffix(self):
+    #     return self.versionsuffix or ''
+
+
+# implement this to your own needs - to create custom yaml/json/xml parser
+class GenericSpecsParser(object):
+    """Parent of all implemented parser classes"""
+    @ staticmethod
+    def parse(filename):
+        """Parent of all implemented parser functions (i.e. JSON, XML...)"""
+        raise NotImplementedError
+
+
+class YamlSpecParser(GenericSpecsParser):
+    """YAML file parser"""
+    @ staticmethod
+    def parse(filename):
+        """Parses YAML file and assigns obtained values to SW config instances as well as general config instance"""
+        try:
+            with open(filename, 'r') as f:
+                spec_dict = yaml.safe_load(f)
+
+            eb = Easystack()
+        except Exception:
+            raise EasyBuildError("Could not read provided easystack.")
+
+        sw_dict = spec_dict["software"]
+
+        # assign software-specific EB attributes
+        for software in sw_dict:
+            try:
+                # iterates through toolchains to find out what sw version is needed
+                for yaml_toolchain in sw_dict[software]['toolchains']:
+                    # retrieves version number
+                    for yaml_version in sw_dict[software]['toolchains'][yaml_toolchain]['versions']:
+
+                        # if among versions there is anything else than known flags or version flag itself, it is wrong
+                        # identification of version strings is vague
+                        if str(yaml_version)[0].isdigit() \
+                                or str(yaml_version)[-1].isdigit():
+                            # creates a sw class instance
+                            try:
+                                yaml_toolchain_name = str(yaml_toolchain).split('-', 1)[0]
+                                yaml_toolchain_version = str(yaml_toolchain).split('-', 1)[1]
+                            except IndexError:
+                                yaml_toolchain_name = str(yaml_toolchain)
+                                yaml_toolchain_version = ''
+
+                            sw = SoftwareSpecs(software=software, version=yaml_version,
+                                               toolchain=yaml_toolchain, toolchain_name=yaml_toolchain_name,
+                                               toolchain_version=yaml_toolchain_version)
+
+                            # append newly created class instance to the list inside EbFromSpecs class
+                            eb.software_list.append(sw)
+
+                        elif str(yaml_version) == 'versionsuffix':
+                            try:
+                                version_info = sw_dict[software]['toolchains'][yaml_toolchain]['versions'][yaml_version]
```

Comment:
This is why you use variables for stuff you use in multiple places. Use `toolchain_data` here instead of `sw_dict[software]['toolchains'][yaml_toolchain]`, that will makes this a lot more readable.